### ISSUE TITLE
Nightly build failures with -Werror=uninitialized: cuda/11.1.0,11.4.2 builds, Kokkos_Combined_Reducer.hpp various is used uninitialized in this function warnings

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#6125]
**Describe the bug**
Nightly cuda/11.1 builds (so far) are failing with -Werror=uninitialized:

```
18:29:23 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830_cpp17_uvm/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult0' is used uninitialized in this function [-Werror=uninitialized]
18:29:23    auto combined_value = Impl::make_combined_reducer_value<mem_space_type>(
18:29:23                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   ^                                         
18:29:23 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830_cpp17_uvm/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult1' is used uninitialized in this function [-Werror=uninitialized]
18:29:23 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830_cpp17_uvm/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult2' is used uninitialized in this function [-Werror=uninitialized]
18:29:23 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830_cpp17_uvm/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult3' is used uninitialized in this function [-Werror=uninitialized]
```

These commits merged since the previous passing build:

Work around NVHPC issue with enum types — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/ea134de4816d0dd0cc052325ffe327c93f755c46)
    Work around NVHPC 23.x issues — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/60b982ad090d7ef6686c61181db0cae6983ca95b)
    Revert "Merge pull request #5964 from PhilMiller/cuda-lambda-default" — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/d7c06c4335c2329c17881f207a4675ea11b8ea6e)
    Always pass -extended-lambda option to NVCC and force Kokkos_ENABLE_CUDA_LAMBDA ON — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/798efc5a38e8e57a4982bc235a301ccd46a12a4d)
    Drop CUDA_LAMBDA guards in Cuda headers — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/0954a1b0a28063211c02b83adb387c19ef2827d1)
    Work around NVHPC 23.x not dealing with __isGlobal — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/417a6ee735d650bd20116a1e999e6854594e8582)
    Reintroduce test skip for nvhpc < 23.3 — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/41d9d06269e2115f6477a80ba07ed14e6fdd647e)
    hpcbind: check for correct Slurm variable — [rberger](https://jenkins-son.sandia.gov/user/rberger/) / [githubweb](https://github.com/kokkos/kokkos/commit/aa7ab5fd59ffed3625eff6e9a0eb11b49d4bc528)
    Fix macro guards in test for NVC++ as the CUDA compiler — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/531b01dce1ca3ef235c1678f7d4875513ce465e5)
    Allow templated functors in parallel_for, parallel_reduce and parallel_scan (#5976) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/c62a42e1c6d83f0b6ab7a27928185a79445665da)
    Qualify calls possibly ambiguous calls to bit_cast — [arndtd](https://jenkins-son.sandia.gov/user/arndtd/) / [githubweb](https://github.com/kokkos/kokkos/commit/5f45c3086b9e33539d15e36971895b5df67fe52d)
    Fix nightlies -- workaround compiler bug in GCC 9.1 and 9.2 (#6118) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/6b2459ce68d257b6dbdb6964b93e39de68d39223)
    Kokkos_BitManipulation: KOKKOS_COMPILER_GCC->KOKKOS_COMPILER_GNU (#6119) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/e94b5dd363977e77f1d7d151eedd83dc7bc3512a)

**Reproducer**:
```bash
module load sems-cmake sems-gcc/8.3.0 sems-cuda/11.1.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="17" --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda,uvm  --no-examples   --with-options=enable_large_mem_tests
```

**Comment by dalg24** [comment]
What target was it building?  Is it failing with the Cuda execution space or some Host execution space?

**Comment by dalg24** [comment]
Also please confirm that you did not get the warning with e247508

**Comment by ndellingwood** [comment]
@dalg24 

* The warning does not occur with https://github.com/kokkos/kokkos/commit/e247508644d8e1262ffb92289c9bc474342ea37a

* This was building with Cuda enabled (Serial as the host backend), here is verbose build output from `make -j1 VERBOSE=1`:
```
cd /ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/core/unit_test && /ascldap/users/ndellin/kokkos/bin/nvcc_wrapper -DKOKKOS_DEPENDENCE -I/ascldap/users/ndellin/kokkos/tpls/gtest -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/core/unit_test -I/ascldap/users/ndellin/kokkos/core/unit_test -I/ascldap/users/ndellin/kokkos/core/unit_test/category_files -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830 -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/core/src -I/ascldap/users/ndellin/kokkos/core/src -I/ascldap/users/ndellin/kokkos/core/src/../../tpls/desul/include -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/containers/src -I/ascldap/users/ndellin/kokkos/containers/src -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/algorithms/src -I/ascldap/users/ndellin/kokkos/algorithms/src -I/ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/simd/src -I/ascldap/users/ndellin/kokkos/simd/src -isystem /projects/sems/install/rhel7-x86_64/sems/v2/tpl/cuda/11.1.0/gcc/8.3.0/base/hriet6f/include -O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -O3 -DNDEBUG -extended-lambda -Wext-lambda-captures-this -arch=sm_70 -std=c++17 -MD -MT core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/serial/TestSerial_TeamCombinedReducers.cpp.o -MF CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/serial/TestSerial_TeamCombinedReducers.cpp.o.d -o CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/serial/TestSerial_TeamCombinedReducers.cpp.o -c /ascldap/users/ndellin/kokkos/testing/Cuda111_Gcc830/core/unit_test/serial/TestSerial_TeamCombinedReducers.cpp
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp: In static member function 'static OpFuncR __nv_hdl_wrapper_t<false, false, Tag, OpFuncR(OpFuncArgs ...), F1, F2>::manager<Lambda>::do_call(void*, OpFuncArgs ...) [with Lambda = _GLOBAL__N__66_tmpxft_000003db_00000000_7_TestSerial_TeamCombinedReducers_cpp1_ii_f98ad614::TeamTeamCombinedReducer::test_team_thread_range_only_scalars(int)::<lambda(const team_member_type&)>; Tag = __nv_dl_tag<void (_GLOBAL__N__66_tmpxft_000003db_00000000_7_TestSerial_TeamCombinedReducers_cpp1_ii_f98ad614::TeamTeamCombinedReducer::*)(int), &_GLOBAL__N__66_tmpxft_000003db_00000000_7_TestSerial_TeamCombinedReducers_cpp1_ii_f98ad614::TeamTeamCombinedReducer::test_team_thread_range_only_scalars, 1>; OpFuncR = void; OpFuncArgs = {const Kokkos::Impl::HostThreadTeamMember<Kokkos::Serial>&}; F1 = const int; F2 = Kokkos::View<int [4], Kokkos::HostSpace>]':
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult0' is used uninitialized in this function [-Werror=uninitialized]
   auto combined_value = Impl::make_combined_reducer_value<mem_space_type>(
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   ^                                         
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult1' is used uninitialized in this function [-Werror=uninitialized]
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult2' is used uninitialized in this function [-Werror=uninitialized]
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult3' is used uninitialized in this function [-Werror=uninitialized]
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp: In lambda function:
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult0' is used uninitialized in this function [-Werror=uninitialized]
   auto combined_value = Impl::make_combined_reducer_value<mem_space_type>(
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   ^                                         
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult3' is used uninitialized in this function [-Werror=uninitialized]
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp: In lambda function:
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult0' is used uninitialized in this function [-Werror=uninitialized]
   auto combined_value = Impl::make_combined_reducer_value<mem_space_type>(
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                   ^                                         
/ascldap/users/ndellin/kokkos/core/src/impl/Kokkos_Combined_Reducer.hpp:508:110: error: 'teamResult3' is used uninitialized in this function [-Werror=uninitialized]
cc1plus: all warnings being treated as errors
make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/serial/TestSerial_TeamCombinedReducers.cpp.o] Error 1
make[2]: Leaving directory `/home/ndellin/kokkos/testing/Cuda111_Gcc830'
make[1]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Serial2.dir/all] Error 2
make[1]: Leaving directory `/home/ndellin/kokkos/testing/Cuda111_Gcc830'
make: *** [all] Error 2
```

**Comment by dalg24** [comment]
> The warning does not occur with https://github.com/kokkos/kokkos/commit/e247508644d8e1262ffb92289c9bc474342ea37a

If I had to guess I would say look whether it is related to #5976 although it is non obvious to me how that would cause the issue.

**Comment by ndellingwood** [comment]
The error also occurred with cuda/11.4.2, I updated the title to include that

**Comment by dalg24** [comment]
Do you get the warning with that commit ? Nathan said no earlier.

**Comment by masterleinad** [comment]
> Do you get the warning with that commit ? Nathan said no earlier.

Yes, I bisected it starting from 4.0.00. The relevant code clearly didn't change in that pull request but I haven't been able to reproduce the warning just using host-side `parallel_reduce` with multiple reducers using uninitialized values yet.

**Comment by ndellingwood** [comment]
@masterleinad #6126 resolved the -Werror for me

When I tested https://github.com/kokkos/kokkos/commit/e247508644d8e1262ffb92289c9bc474342ea37a I did not encounter the -Werror , following the reproducer I posted

**Comment by ndellingwood** [comment]
As a sanity check I just retested with https://github.com/kokkos/kokkos/commit/e247508644d8e1262ffb92289c9bc474342ea37a in a clean build and did not get hit with any -Werror, the build compiled successfully. Though that is the sha I would have expected to introduce the associated -Werror as well

**Comment by ndellingwood** [comment]
@dalg24 I'll take a look at that later, the Tue nightly build passed and then failed last night (Wed), for reference these are the commits that went in from the previous pass:

Work around NVHPC issue with enum types — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/ea134de4816d0dd0cc052325ffe327c93f755c46)
Work around NVHPC 23.x issues — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/60b982ad090d7ef6686c61181db0cae6983ca95b)
Revert "Merge pull request https://github.com/kokkos/kokkos/pull/5964 from PhilMiller/cuda-lambda-default" — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/d7c06c4335c2329c17881f207a4675ea11b8ea6e)
Always pass -extended-lambda option to NVCC and force Kokkos_ENABLE_CUDA_LAMBDA ON — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/798efc5a38e8e57a4982bc235a301ccd46a12a4d)
Drop CUDA_LAMBDA guards in Cuda headers — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/0954a1b0a28063211c02b83adb387c19ef2827d1)
Work around NVHPC 23.x not dealing with __isGlobal — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/417a6ee735d650bd20116a1e999e6854594e8582)
Reintroduce test skip for nvhpc < 23.3 — [Christian Robert Trott](https://jenkins-son.sandia.gov/user/crtrott/) / [githubweb](https://github.com/kokkos/kokkos/commit/41d9d06269e2115f6477a80ba07ed14e6fdd647e)
hpcbind: check for correct Slurm variable — [rberger](https://jenkins-son.sandia.gov/user/rberger/) / [githubweb](https://github.com/kokkos/kokkos/commit/aa7ab5fd59ffed3625eff6e9a0eb11b49d4bc528)
Fix macro guards in test for NVC++ as the CUDA compiler — [dalg24](https://jenkins-son.sandia.gov/user/dalg24/) / [githubweb](https://github.com/kokkos/kokkos/commit/531b01dce1ca3ef235c1678f7d4875513ce465e5)
Allow templated functors in parallel_for, parallel_reduce and parallel_scan (https://github.com/kokkos/kokkos/pull/5976) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/c62a42e1c6d83f0b6ab7a27928185a79445665da)
Qualify calls possibly ambiguous calls to bit_cast — [arndtd](https://jenkins-son.sandia.gov/user/arndtd/) / [githubweb](https://github.com/kokkos/kokkos/commit/5f45c3086b9e33539d15e36971895b5df67fe52d)
Fix nightlies -- workaround compiler bug in GCC 9.1 and 9.2 (https://github.com/kokkos/kokkos/pull/6118) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/6b2459ce68d257b6dbdb6964b93e39de68d39223)
Kokkos_BitManipulation: KOKKOS_COMPILER_GCC->KOKKOS_COMPILER_GNU (https://github.com/kokkos/kokkos/pull/6119) — [github](https://jenkins-son.sandia.gov/user/github/) / [githubweb](https://github.com/kokkos/kokkos/commit/e94b5dd363977e77f1d7d151eedd83dc7bc3512a)

**Comment by masterleinad** [comment]
The errors are originating, e.g., from https://github.com/kokkos/kokkos/blob/e94b5dd363977e77f1d7d151eedd83dc7bc3512a/core/unit_test/TestTeamCombinedReducers.hpp#L34-L56 where we don't initialize `teamResult0`, `teamResult1`, `teamResult2`, `teamResult3`.

**Comment by masterleinad** [comment]
The behavior of `_get_value_from_combined_reducer_ctor_arg` hasn't changed for non-reducer, non-View objects since it was introduced in https://github.com/kokkos/kokkos/commit/624b0a6b513d8c8d50c2d82226dbb0c3c646cec5.

**Comment by ndellingwood** [comment]
Resolved by #6126

**Comment by masterleinad** [comment]
Fixed in #6126.
